patches/master_patch.patch

```diff
@@ -1831,6 +1831,488 @@ index a433afd0217830b49ddb3381ad55afeae2381840..ddcca749666f956c6317676dbcdd48da
++++ b/net/proxy/proxy_server.cc
+@@ -65,7 +65,23 @@ ProxyServer::Scheme GetSchemeFromURIInternal(base::StringPiece type) {
+ }  // namespace
+ 
+ ProxyServer::ProxyServer(Scheme scheme, const HostPortPair& host_port_pair)
+-      : scheme_(scheme), host_port_pair_(host_port_pair) {
++      : scheme_(scheme), host_port_pair_(host_port_pair), auth_(false) {
++  if (scheme_ == SCHEME_DIRECT || scheme_ == SCHEME_INVALID) {
++    // |host_port_pair| isn't relevant for these special schemes, so none should
++    // have been specified. It is important for this to be consistent since we
++    // do raw field comparisons in the equality and comparison functions.
++    DCHECK(host_port_pair.Equals(HostPortPair()));
++    host_port_pair_ = HostPortPair();
++  }
++}
++
++ProxyServer::ProxyServer(Scheme scheme, const HostPortPair& host_port_pair,
++                         const std::string& username,
++                         const std::string& password)
++      : scheme_(scheme), host_port_pair_(host_port_pair),
++        auth_(true), username_(username), password_(password) {
++  // Authentication is supported only for SOCKS5 proxies.
++  DCHECK(scheme_ == SCHEME_SOCKS5);
+   if (scheme_ == SCHEME_DIRECT || scheme_ == SCHEME_INVALID) {
+     // |host_port_pair| isn't relevant for these special schemes, so none should
+     // have been specified. It is important for this to be consistent since we
+@@ -83,6 +99,16 @@ const HostPortPair& ProxyServer::host_port_pair() const {
+   return host_port_pair_;
+ }
+ 
++const std::string& ProxyServer::username() const {
++  DCHECK(is_auth());
++  return username_;
++};
++
++const std::string& ProxyServer::password() const {
++  DCHECK(is_auth());
++  return password_;
++};
++
+ // static
+ ProxyServer ProxyServer::FromURI(const std::string& uri,
+                                  Scheme default_scheme) {
+@@ -123,7 +149,8 @@ std::string ProxyServer::ToURI() const {
+     case SCHEME_SOCKS4:
+       return std::string("socks4://") + host_port_pair().ToString();
+     case SCHEME_SOCKS5:
+-      return std::string("socks5://") + host_port_pair().ToString();
++      return std::string("socks5://") + AuthString() +
++        host_port_pair().ToString();
+     case SCHEME_HTTPS:
+       return std::string("https://") + host_port_pair().ToString();
+     case SCHEME_QUIC:
+@@ -175,7 +202,8 @@ std::string ProxyServer::ToPacString() const {
+       // For compatibility send SOCKS instead of SOCKS4.
+       return std::string("SOCKS ") + host_port_pair().ToString();
+     case SCHEME_SOCKS5:
+-      return std::string("SOCKS5 ") + host_port_pair().ToString();
++      return std::string("SOCKS5 ") + AuthString() +
++        host_port_pair().ToString();
+     case SCHEME_HTTPS:
+       return std::string("HTTPS ") + host_port_pair().ToString();
+     case SCHEME_QUIC:
+@@ -187,6 +215,16 @@ std::string ProxyServer::ToPacString() const {
+   }
+ }
+ 
++std::string ProxyServer::AuthString() const {
++  if (!is_auth())
```

Comment:
I think we can leave out the whole `is_auth()` thing. If the username/password are both empty then we assume no auth

Reply:
actually it seems like this whole thing could be simplified by adding username/password support to HostPortPair